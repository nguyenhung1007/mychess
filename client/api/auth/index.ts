import http from '../axiosCommon';
import { AxiosInstance } from 'axios';
import { UserLoginDto, UserRegisterDto, ForgotPasswordEmailDto, ForgotPasswordPhoneDto } from './dto';
import { ApiResponse } from '../../store/api/interface';

export class AuthAPI {
    constructor(private readonly apiCall: AxiosInstance, readonly prefix: string) {}

    async loginUser(input: UserLoginDto) {
        const url = `${this.prefix + '/login'}`;
        const res = await this.apiCall.post<ApiResponse<null>>(url, input);
        return res;
    }

    async getSocketToken() {
        const url = `${this.prefix + '/socket-token'}`;
        const res = await this.apiCall.get<ApiResponse<null>>(url);
        return res;
    }

    async logoutUser() {
        const url = `${this.prefix + '/logout'}`;
        const res = await this.apiCall.post<ApiResponse<null>>(url);
        return res;
    }

    async registerUser(input: UserRegisterDto) {
        const url = `${this.prefix + '/register'}`;
        const res = await this.apiCall.post<ApiResponse<null>>(url, input);
        return res;
    }

    async forgotPasswordByEmail(input: ForgotPasswordEmailDto) {
        const url = `${this.prefix + '/otp-email'}`;
        const res = await this.apiCall.post<ApiResponse<void>>(url, input);
        return res;
    }

    async forgotPasswordByPhone(input: ForgotPasswordPhoneDto) {
        const url = `${this.prefix + '/otp-sms'}`;
        const res = await this.apiCall.post<ApiResponse<void>>(url, input);
        return res;
    }
    async checkOTP(input: string) {
        const url = `${this.prefix + '/check-otp?key=' + input}`;
        const res = await this.apiCall.post<ApiResponse<void>>(url);
        return res;
    }
}
export const authApi = new AuthAPI(http, '/auth');
export default authApi;
