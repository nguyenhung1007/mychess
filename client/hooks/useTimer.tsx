import { useEffect, useState } from "react";

export interface UseTimerProps {
        resetTime: number;
        defaultStop?: boolean;
}

const UseTimer = ({ resetTime, defaultStop = false }: UseTimerProps): [number, boolean, () => void] => {
        const [startTime, setStartTime] = useState<number>(new Date().getTime());
        const [endTime, setEndTime] = useState(new Date().getTime());
        const [isFinish, setFinish] = useState(defaultStop);

        useEffect(() => {
                if (!isFinish) {
                        setStartTime(new Date().getTime());
                }
        }, [isFinish]);

        useEffect(() => {
                if (!isFinish) {
                        const countDownEvent = setInterval(() => {
                                if (endTime - startTime > resetTime) {
                                        setFinish(true);
                                } else setEndTime(new Date().getTime());
                        }, 1000);

                        return () => {
                                clearInterval(countDownEvent);
                        };
                }
        }, [startTime, endTime, isFinish]);

        return [endTime - startTime < 0 ? 0 : Math.round((endTime - startTime) / 1000), isFinish, () => setFinish(false)];
};

export default UseTimer;
