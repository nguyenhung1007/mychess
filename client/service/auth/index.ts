import axios, { AxiosStatic } from "axios";
import { ApiResponse } from "../../store/api/dto";
import { RegisterUserDto } from "./dto";

class AuthService {
        private endPoint: string = "/auth";
        constructor(private http: AxiosStatic) {}

        public async registerUser(input: RegisterUserDto) {
                return await this.http.post<ApiResponse<null>>(this.endPoint + "/register", { ...input });
        }
}

export const authService = new AuthService(axios);
