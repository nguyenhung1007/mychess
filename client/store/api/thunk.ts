import { createAsyncThunk } from '@reduxjs/toolkit';

import { SupportDto } from '../../common/interface/dto/common.dto';
import { CommonAPI, commonAPI } from '../../api/commonApi';

class CommonThunk {
    constructor(private readonly apiCall: CommonAPI) {}

    sendSupport = createAsyncThunk<null, SupportDto>('sendSupport', async (input) => {
        await this.apiCall.sendFeedBack(input);
        return null;
    });
}
export const commonThunk = new CommonThunk(commonAPI);
export default commonThunk;
