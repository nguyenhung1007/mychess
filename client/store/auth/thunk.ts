import { createAsyncThunk } from '@reduxjs/toolkit';

import { UserLoginDto, UserRegisterDto, ForgotPasswordEmailDto, ForgotPasswordPhoneDto } from '../../api/auth/dto';
import { AuthAPI, authApi } from '../../api/auth';
import { ApiResponse } from '../api/interface';

class AuthThunk {
    constructor(private readonly apiCall: AuthAPI) {}

    loginUser = createAsyncThunk<null, UserLoginDto>('UserLoginDto', async (input) => {
        await this.apiCall.loginUser(input);
        return null;
    });

    logoutUser = createAsyncThunk<null, void>('LogoutUser', async () => {
        await this.apiCall.logoutUser();
        return null;
    });

    getSocketToken = createAsyncThunk<null, void>('getSocketToken', async () => {
        await this.apiCall.getSocketToken();
        return null;
    });

    registerUser = createAsyncThunk<null, UserRegisterDto>('UserRegisterDto', async (input) => {
        await this.apiCall.registerUser(input);
        return null;
    });

    forgotPasswordByEmail = createAsyncThunk<ApiResponse<void>, ForgotPasswordEmailDto>('ForgotPasswordEmailDto', async (input) => {
        const res = await this.apiCall.forgotPasswordByEmail(input);
        return res.data;
    });

    forgotPasswordByPhone = createAsyncThunk<ApiResponse<void>, ForgotPasswordPhoneDto>('ForgotPasswordPhoneDto', async (input) => {
        const res = await this.apiCall.forgotPasswordByPhone(input);
        return res.data;
    });
}
export const authThunk = new AuthThunk(authApi);
export default authThunk;
