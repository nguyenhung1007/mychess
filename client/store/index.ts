import { combineReducers, configureStore } from '@reduxjs/toolkit';
import { apiReducer } from './api';
import { IApiState } from './api/interface';
import { authReducer } from './auth';
import { IAuthState } from './auth/interface';

export interface RootState {
        api: IApiState;
        auth: IAuthState;
}

const reducers = combineReducers<RootState>({
        api: apiReducer,
        auth: authReducer,
});

export const store = configureStore({
        reducer: reducers,
        devTools: process.env.NODE_ENV === 'production',
});
