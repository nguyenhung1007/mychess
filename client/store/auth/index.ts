import { createSlice } from '@reduxjs/toolkit';
import Cookies from 'universal-cookie';

import { AuthState } from './interface';
import { authThunk } from './thunk';
import { userThunk } from './userThunk';

const initialState: AuthState = {
    email: '',
    username: '',
    name: '',
    elo: 0,
    id: '',
    phoneNumber: '',
    avatarUrl: '',
    createDate: Date(),
    isLogin: false,
};

const reducer = createSlice({
    name: 'auth',
    initialState,
    reducers: {
        resetState: () => ({ ...initialState }),
        updateLogin: (state) => ({ ...state, isLogin: true }),
    },
    extraReducers: (builder) => {
        builder.addCase(userThunk.getCurrentUser.fulfilled, (state, { payload }) => {
            const newState = { ...state };
            newState.name = payload.name;
            newState.username = payload.username;
            newState.phoneNumber = payload.phoneNumber;
            newState.email = payload.email;
            newState.elo = payload.elo;
            newState.id = payload.id;
            newState.avatarUrl = payload.avatarUrl;
            newState.isLogin = true;
            return newState;
        });
        builder.addCase(authThunk.loginUser.fulfilled, (state) => ({ ...state, isLogin: true }));
        builder.addCase(authThunk.registerUser.fulfilled, (state) => ({ ...state, isLogin: true }));
        builder.addCase(authThunk.logoutUser.fulfilled, () => ({ ...initialState }));

        builder.addCase(userThunk.getCurrentUser.rejected, (state) => {
            const cookies = new Cookies();
            // cookies.remove('re-token');
            // cookies.remove('auth-token');

            return {
                ...state,
                isLogin: false,
            };
        });
    },
});
export const authActions = {
    ...reducer.actions,
};
export const authReducer = reducer.reducer;
