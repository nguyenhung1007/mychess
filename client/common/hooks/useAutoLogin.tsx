import { useEffect } from 'react';
import { useSelector } from 'react-redux';
import Cookies from 'universal-cookie';
import userApi from '../../api/user';
import { RootState, store } from '../../store';
import { authActions } from '../../store/auth';
import { IAuthState } from '../../store/auth/interface';

export function useAutoLogin() {
    const cookies = new Cookies();
    const authState = useSelector<RootState, IAuthState>((state) => state.auth);

    useEffect(() => {
        const reToken = cookies.get('re-token');
        if (reToken) store.dispatch(authActions.updateLogin());
    }, []);

    useEffect(() => {
        if (authState.isLogin) store.dispatch(userApi.getLoginUser());
    }, [authState.isLogin]);
}

export default useAutoLogin;
