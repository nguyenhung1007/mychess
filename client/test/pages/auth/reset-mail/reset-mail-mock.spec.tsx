import { useState } from 'react';
import * as React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import { Provider, useSelector } from 'react-redux';

import { useFormError } from '../../../../common/hooks/useFormError';
import * as useTimer from '../../../../common/hooks/useTimer';
import ResetEmail from '../../../../pages/auth/reset-mail';
import { store } from '../../../../store';
import { authThunk } from '../../../../store/auth/thunk';

jest.mock('../../../../common/hooks/useFormError');
jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useSelector: jest.fn(),
}));

describe('Register Mock', () => {
    const useFormErrorMock = useFormError as jest.Mock<any, any>;
    const useSelectorMock = useSelector as jest.Mock<any, any>;

    beforeEach(() => {
        useSelectorMock.mockReturnValue({
            isLoading: false,
            errorDetails: {},
            isError: false,
            message: '',
        });
        useFormErrorMock.mockReturnValue({ email: '' });
    });

    afterEach(() => {
        useFormErrorMock.mockClear();
        useSelectorMock.mockClear();
    });

    it('Render Correct Default Mock', async () => {
        let wrapper = render(
            <Provider store={store}>
                <ResetEmail />
            </Provider>,
        );

        const emailTextField = await wrapper.findByTestId('textfield-input-email');

        expect(emailTextField).toBeDefined();

        expect(wrapper).toMatchSnapshot();
        try {
            await wrapper.findByTestId('textfield-error-email');
        } catch (err) {
            expect(err).toBeDefined();
        }

        try {
            await wrapper.findByTestId('waveloading');
        } catch (err) {
            expect(err).toBeDefined();
        }
        try {
            await wrapper.findByTestId('successmsg');
        } catch (err) {
            expect(err).toBeDefined();
        }
    });

    it('Render with error label', async () => {
        useFormErrorMock.mockReturnValue({
            email: 'error-message',
        });

        const wrapper = render(
            <Provider store={store}>
                <ResetEmail />
            </Provider>,
        );

        const emailError = await wrapper.findByTestId('textfield-error-email');

        expect(emailError.innerHTML).toContain('Email error-message');
    });

    it('Render with loading ', async () => {
        useSelectorMock.mockReturnValue({ isLoading: true, errorDetails: {}, isError: false, message: '' });

        const wrapper = render(
            <Provider store={store}>
                <ResetEmail />
            </Provider>,
        );

        const loadingWave = await wrapper.findByTestId('waveloading');
        expect(loadingWave).toBeDefined();
    });

    it('Render with success message ', async () => {
        useSelectorMock.mockReturnValue({ isLoading: false, errorDetails: {}, isError: false, message: 'hello' });

        const wrapper = render(
            <Provider store={store}>
                <ResetEmail />
            </Provider>,
        );

        const successMsg = await wrapper.findByText('hello');
        expect(successMsg).toBeDefined();
    });
    it('Render after submit screen allow resend ', async () => {
        useSelectorMock.mockReturnValue({ isLoading: false, errorDetails: {}, isError: false, message: '' });
        const useStateSpy = jest.spyOn(React, 'useState');
        const useTimerSpy = jest.spyOn(useTimer, 'useTimer');
        useStateSpy.mockImplementation(() => [true, jest.fn()]);
        useTimerSpy.mockImplementation(() => [60, true, jest.fn()]);
        const wrapper = render(
            <Provider store={store}>
                <ResetEmail />
            </Provider>,
        );

        const btnChangeMail = await wrapper.findByTestId('btnform-change-another-email');
        const timerCountDown = await wrapper.findByText('60s');
        const labelSendOther = await wrapper.findByText('Send me an another email.');
        expect(btnChangeMail).toBeDefined();
        expect(timerCountDown).toBeDefined();
        expect(labelSendOther).toBeDefined();

        useStateSpy.mockClear();
    });

    it('Render after submit screen allow resend with client here ', async () => {
        useSelectorMock.mockReturnValue({ isLoading: false, errorDetails: {}, isError: false, message: '' });
        const useStateSpy = jest.spyOn(React, 'useState');
        const useTimerSpy = jest.spyOn(useTimer, 'useTimer');
        useStateSpy.mockImplementation(() => [true, jest.fn()]);
        useTimerSpy.mockImplementation(() => [0, false, jest.fn()]);
        const wrapper = render(
            <Provider store={store}>
                <ResetEmail />
            </Provider>,
        );

        const btnChangeMail = await wrapper.findByTestId('btnform-change-another-email');
        const clickHereBtn = await wrapper.findByText('Click Here');
        const labelSendOther = await wrapper.findByText('Send me an another email.');
        expect(btnChangeMail).toBeDefined();
        expect(clickHereBtn).toBeDefined();
        expect(labelSendOther).toBeDefined();
        useTimerSpy.mockClear();
        useStateSpy.mockClear();
    });

    it('Function submit with isSubmit false', async () => {
        const submitMockFn = jest.fn();
        const authThunkSpy = jest.spyOn(authThunk, 'forgotPasswordByEmail');
        authThunkSpy.mockReturnValue(submitMockFn);

        const wrapper = render(
            <Provider store={store}>
                <ResetEmail />
            </Provider>,
        );
        const formLogin = await wrapper.findByTestId('reset-mail');

        await fireEvent.submit(formLogin);

        expect(submitMockFn).toBeCalledTimes(1);
        authThunkSpy.mockClear();
    });

    it('Function submit with isSubmit True', async () => {
        const authThunkSpy = jest.spyOn(authThunk, 'forgotPasswordByEmail');
        const submitMockFn = jest.fn();
        const useStateSpy = jest.spyOn(React, 'useState');
        useStateSpy.mockImplementation(() => [true, jest.fn()]);

        authThunkSpy.mockReturnValue(submitMockFn);

        const wrapper = render(
            <Provider store={store}>
                <ResetEmail />
            </Provider>,
        );
        const formLogin = await wrapper.findByTestId('reset-mail');
        await fireEvent.submit(formLogin);
        expect(submitMockFn).toBeCalledTimes(1);

        authThunkSpy.mockClear();
        useStateSpy.mockClear();
    });
});
