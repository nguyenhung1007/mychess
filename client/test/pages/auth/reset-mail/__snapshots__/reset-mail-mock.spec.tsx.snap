// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Mock Render Correct Default Mock 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex-1 chess-bg grid place-items-center grid-rows-max shadow-sm"
      >
        <div
          class="bg-gray-800 px-4 md:px-10 py-12 w-full max-w-md rounded-sm fade-in "
        >
          <form
            data-testid="reset-mail"
          >
            <h1
              class="text-center text-4xl text-white mb-7"
            >
              Reset Password
            </h1>
            <p
              class="text-mercury-800 py-2"
            >
              Please enter your email, you will receive an mail to reset your password
            </p>
            <div
              class="space-y-2"
            >
              <div
                class="space-y-1.5 text-sm"
              >
                <label
                  class="block text-cotton-seed "
                  data-testid="textfield-label-email"
                  for="email"
                >
                  Email
                </label>
                <input
                  autocomplete="on"
                  class="block w-full outline-none rounded-sm bg-tuna py-2 px-1.5  text-mercury"
                  data-testid="textfield-input-email"
                  id="email"
                  name="email"
                  type="text"
                />
              </div>
            </div>
            <div
              class="mt-4 mb-4"
            >
              <button
                class="bg-btn-1 w-full py-2 text-white rounded-sm focus:outline-none "
                data-testid="btnform-send-an-email"
                type="submit"
              >
                Send An Email
              </button>
            </div>
          </form>
          <div>
            <div
              class="text-center"
              data-testid="sidelink-wrapper-or-reset-by-phone-number"
            >
              <a
                class="text-mercury text-sm duration-300 hover:text-malibu"
                data-testid="sidelink-link-or-reset-by-phone-number"
                href="/auth/reset-phone"
              >
                Or Reset By Phone Number?
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex-1 chess-bg grid place-items-center grid-rows-max shadow-sm"
    >
      <div
        class="bg-gray-800 px-4 md:px-10 py-12 w-full max-w-md rounded-sm fade-in "
      >
        <form
          data-testid="reset-mail"
        >
          <h1
            class="text-center text-4xl text-white mb-7"
          >
            Reset Password
          </h1>
          <p
            class="text-mercury-800 py-2"
          >
            Please enter your email, you will receive an mail to reset your password
          </p>
          <div
            class="space-y-2"
          >
            <div
              class="space-y-1.5 text-sm"
            >
              <label
                class="block text-cotton-seed "
                data-testid="textfield-label-email"
                for="email"
              >
                Email
              </label>
              <input
                autocomplete="on"
                class="block w-full outline-none rounded-sm bg-tuna py-2 px-1.5  text-mercury"
                data-testid="textfield-input-email"
                id="email"
                name="email"
                type="text"
              />
            </div>
          </div>
          <div
            class="mt-4 mb-4"
          >
            <button
              class="bg-btn-1 w-full py-2 text-white rounded-sm focus:outline-none "
              data-testid="btnform-send-an-email"
              type="submit"
            >
              Send An Email
            </button>
          </div>
        </form>
        <div>
          <div
            class="text-center"
            data-testid="sidelink-wrapper-or-reset-by-phone-number"
          >
            <a
              class="text-mercury text-sm duration-300 hover:text-malibu"
              data-testid="sidelink-link-or-reset-by-phone-number"
              href="/auth/reset-phone"
            >
              Or Reset By Phone Number?
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
